crosshair/libimpl/builtinslib.py

```diff
@@ -1967,6 +1990,24 @@ def split(self, sep: Optional[str] = None, maxsplit: int = -1):
         ret.extend(self[first_occurance + 1 :].split(sep=sep, maxsplit=new_maxsplit))
         return ret
 
+    def rsplit(self, sep: Optional[str] = None, maxsplit: int = -2):
+        if sep is None:
+            return self.__str__().split(sep=sep, maxsplit=maxsplit)
+        smt_sep = force_to_smt_sort(sep, SmtStr)
+        if not isinstance(maxsplit, Integral):
+            raise TypeError
+        if maxsplit == 0:
+            return [self]
+        last_occurence = self.rfind(sep)
+        if last_occurence == -1:
+            return [self]
+        new_maxsplit = -1 if maxsplit == -1 else maxsplit - 1
+        remaining = self[: cast(int, last_occurence)]
+        ret = self[:last_occurence].rsplit(sep, new_maxsplit)
+        index_after = SmtInt(z3.Length(smt_sep)) + last_occurence
```

Comment:
```suggestion
        index_after = len(sep) + last_occurence
```
Equivalent (I believe). Just a tad simpler. (and it doesn't use z3 in case `sep` is a concrete string)

Reply:
Yeah I ended up at this construction by fighting with the error messages for half an hour. Still need more solid intuition for what happens in z3-land and what happens in Python-land and how `Smt` brokers between those.

Reply:
This is tricky - there are three different kinds of values you might have for any python type. Check out [the table here](https://crosshair.readthedocs.io/en/latest/how_does_it_work.html#the-basic-idea) if you haven't already.

It's worth noting that the Smt* classes emulate native python values, so if you know you have those, you can just use regular python on them. 

There are a few different ways to get a z3 value. If you know you've got an Smt* instance, you can just access the `.var` attribute of it. The `coerce_to_smt_sort(...)` / `force_to_smt_sort(...)` functions are helpers that can take either Smt* instances or native Python values and give you back the appropriate z3 value.

When going the other way, you just construct the Smt*() instance and pass in the z3 value.

Not sure whether any of this helps or even makes any sense!

Reply:
Ah so where my thinking was going funky was that I assumed `coerce_to_smt_sort(...)` would give me an Smt-object, not a z3-object!